### PR TITLE
Fixed LOH byte array allocations in the IDE from metadata

### PR DESCRIPTION
**Pull request by TIHan**
When doing a find all references in VS on a `bool` type in the VisualFSharp.sln, byte arrays are the top LOH allocations, mainly around FSharp metadata. 

This PR should resolve some of it by using memory mapped files. It also unifies the memory handling of both IL and FSharp metadata with the `ByteMemory` type. 

A `ByteMemory` can be backed by in-memory or memory mapped file data.

Let's see if CI passes. :)

### CODE REVIEW

**Comment by TIHan** [comment]
CI failed not as a result of this change, it's due to this commit: https://github.com/dotnet/fsharp/commit/d68942f26a6c9c1399e2760dace4b7646c274875

**Comment by KevinRansom** [comment]
Will #7970 Should fix the CI.  It looks to me like the failures here are fixed by #7970

**Comment by TIHan** [comment]
Great! Thank you!

**Comment by TIHan** [comment]
This is ready.

**Comment by TIHan** [comment]
I prefer interfaces if everything was self contained in the file and nothing else would use it. Since we are using `ByteMemory` throughout the compiler, I don't want to use an interface because I don't want some other component to implement their own `IByteMemory` and use it.

File: src/absil/bytes.fs [hunk 1]
```diff
@@ -26,10 +33,290 @@ module internal Bytes =
...
+[<Sealed>]
+type ByteArrayMemory(bytes: byte[], offset, length) =
+    inherit ByteMemory()
+
+    do
+        if offset < 0 then
+            failwith "offset is less than zero"
+
+        if length <= 0 then
+            failwith "length is less than or equal to zero"
```
Comment (cartermp) [inline src/absil/bytes.fs]: More specific exception type: `ArgumentOutOfRangeException`

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/using-standard-exception-types#exception-and-systemexception

File: src/absil/bytes.fs [hunk 2]
```diff
@@ -26,10 +33,290 @@ module internal Bytes =
...
+
+        if (offset + length) > bytes.Length then
+            failwith "span larger than byte array"
```
Comment (cartermp) [inline src/absil/bytes.fs]: `failwithf` and format the arg names

**Comment by TIHan** [comment]
Gotcha, will do. Thanks for the feedback!

**Comment by TIHan** [comment]
@cartermp I think this is ready with the feedback changes.

**Pull request merged**